### ISSUE TITLE
Odd spacing in show-off document

### ISSUE THREAD

**Issue by haberman** [sile-typesetter/sile#623]
I love the ideas behind SILE so I was trying to evaluate its core typesetting features. When I looked at the [show-off document linked from the readme](https://rawgit.com/simoncozens/sile/b66b979a6dca0c60bd4aa2cbad3da36ae2073672/examples/showoff.pdf) I noticed it has a title that looks squished. It almost looks like there aren't any spaces between the words.

<img width="752" alt="Screen Shot 2019-05-30 at 9 34 02 AM" src="https://user-images.githubusercontent.com/1270/58648453-96635480-82be-11e9-8d54-5ab42868606a.png">

I tried running Sile on my local machine (SILE 0.9.5.1 on macOS, installed from Homebrew) and I got better results for the title spacing, but the space between "Show-off" and "Document" looks a little too big compared to the other spaces. And there are odd spaces in the paragraph, like after "InDesign" and before "Let's". Also the title doesn't appear to be properly centered relative to either the page or the text boxes below (it is too far to the right), but maybe this is an intended effect of how the frames are set up, I can't tell.

<img width="1239" alt="Screen Shot 2019-05-30 at 9 46 13 AM" src="https://user-images.githubusercontent.com/1270/58649153-366dad80-82c0-11e9-83ed-69abf979e5de.png">

Are these spacing problems a bug? Can this be improved somehow?

**Comment by simoncozens** [comment]
Thanks - I think the bug must be with the version we have linked to in the README, since it's clearly building correctly now (and with the additional third page). And I'm not entirely sure why I removed `showoff.pdf` from the repo. That said I know @alerque is not happy about me throwing binary blobs into the repo so I'm not planning to put it back in again right now. Maybe we need a better way to host (built) examples?

**Comment by DavidLRowe** [comment]
@simoncozens Perhaps create a GitHub release (https://github.blog/2013-07-02-release-your-software/) and include the .pdf examples and documentation at that point? "You can also attach binary assets (such as compiled executables, minified scripts, documentation) to a release."

**Comment by haberman** [comment]
@simoncozens Are you saying that the output in my second example is correct? There are still big spaces in the paragraph and the spacing of the title still seems off.

**Comment by alerque** [comment]
The spaces in the title is probably an iteration of #261 which has long since been fixed, but Simon is right I've been whining a lot recently because updating raster images and binary blobs in the repo every time we fix a typesetter bug is a bad idea. Attaching the latest manual and maybe a zip of examples to releases is not a bad idea actually. That can be done in parallel with other solutions too. I'll make a note of that for the release process (see #588).

@haberman The other spacing things you mention are some combination of bad input and bugs. For a show-off document it shows of an awful lot of bad implementations. In particular the `linebreak.tolerance` override in there both should not be necessary and breaks something else (unreasonable amounts of stretch inside lines). That should probably be ~~opened as it's own bug~~ considered the main issue in this issue report. See also #620 for an issue about what happens without that value being set.

**Comment by haberman** [comment]
@alerque Thanks for taking this on!

What do you think about the spacing between "Show-off" and "Document"? To my eye that space looks too big compared to the other spaces. I'm a type novice but should kerning take into account the shape of the "f" and slide "Document" left a bit?

**Comment by alerque** [comment]
The kerning you are referring to is not SILE's responsibility. That is up to the font itself and the shaper. I agree it doesn't look great but in this case I suspect the font just doesn't have good kerning set for the `ff` ligature.

You _can_ manually adjust the size of the space skip and override the one set by the font, and you _could_ manually kern that location, but automatically adjusting this based on the shape of the glyph is beyond the scope of SILE.

**Comment by simoncozens** [comment]
*TLDR: Bad font and we should use a nicer one.*

There's weirder stuff going on there. First, we don't use the ligature because it's discretionary. Baskerville *adds* nearly 300 units of kerning between `f` and `space`, but it doesn't add any kerning between `f_f` and `space`. This means that words ending in `f` do seem to have a lot more space than other words. I don't know why they've done that. This looked fine in the past because we didn't implement space kerning, but measured the space character independently instead. Now we do, it's exposing Baskerville's kerning weirdness.

**Comment by haberman** [comment]
Ah, yes I can confirm what @simoncozens says. Here is a screenshot from Apple Pages. The second line uses the ligature and has much more reasonable kerning.

<img width="819" alt="Screen Shot 2019-06-01 at 9 20 51 AM" src="https://user-images.githubusercontent.com/1270/58751149-a8c0c800-844f-11e9-9574-bb2f7e987d2b.png">

I didn't realize the font was in charge of spacing and kerning. I thought the box-and-glue layout engines of TeX and SILE places individual characters based on the stretch of each line? I guess I have some more reading to do, this looks like a good start. :) https://simoncozens.github.io/fonts-and-layout/features.html

**Comment by alerque** [comment]
I just realized trying to confirm Simon's comments about ligatures (I was going to try turning them on in the show-off doc) that I don't even have Baskerville on my system, I have Impallari's Libre Baskerville and some system level aliases to make it use that instead. In this case I get ligatures by default and sane kerning out of the box ;-) 

Here's what I get  from SILE right now:

![image](https://user-images.githubusercontent.com/173595/58752193-a09a7400-84b3-11e9-83a9-cae69b947fb5.png)

With the exception of the `T`, I'd say the knock off font is better. We still may want to pick a different font, but...

**Comment by simoncozens** [comment]
> I didn't realize the font was in charge of spacing and kerning. I thought the box-and-glue layout engines of TeX and SILE places individual characters based on the stretch of each line? I guess I have some more reading to do, this looks like a good start.

Well, historically yes. We measured the space in the font once and used that everywhere. But then Martin pointed out that for Nastaliq you want to vary the size of the space depending on the surrounding context, so I added an (undocumented) setting which asks the shaper for the width of the space. Oddly enough I was just adding the documentation that setting this afternoon...

> There are multiple ways of defining the space between words. By default, the space between words is determined by the width of the space character in the current font. To help with justifying the text, the spaces are shrinkable and stretchable. Specifically, if the width of a space in the current font settings is `<space>`, then the width of the space between words is `<shaper.spaceenlargementfactor> * <space> plus <shaper.spacestretchfactor> * <space> minus <shaper.spaceshrinkfactor> * <space>`. The default values of these settings make the space width `1.2 <space> plus 0.5 <space> minus 0.333 <space>`.
> 
> If you want to set the word space width explicitly, you can set the `document.spaceskip` setting. You will also need to turn *off* the setting `shaper.variablespaces`, which allows the width of a space to vary based on context (otherwise known as “space kerning”). If you want to go back to the default (measuring the space character of the font), then you need to turn on `shaper.variablespaces` (set it to a true value) and also *unset* the setting `document.spaceskip`. To unset it, just call `\set` with no `value` parameter: `\\set[parameter=document.spaceskip]`.

**Comment by pkulchenko** [comment]
@simoncozens, @alerque, given a [recent discussion](https://news.ycombinator.com/item?id=22393857) of similar issues in the manual, should it be updated to fix the numbers in the TOC, image wrapping on page 4, and large space on page 15? I'm not sure what the problem with the TOC numbers is, but the ShowOff document doesn't exhibit it.

**Comment by alerque** [comment]
The PR #624 is going to close this issue soon. Thanks for reporting it @haberman.

Addressed points are:

* The headline font has been replaced with a free alternative that doesn't have the spacing issue shown.

* The headline has been actually centered (previously paragraph indent was getting added to the leading edge, see #625)

* As @DavidLRowe [suggested](https://github.com/sile-typesetter/sile/issues/623#issuecomment-497418092) the release process has been automated and generated PDFs are now automatically updated and attached to releases (see #588 and many others). Currently only the manual is being posted on releases, but all the examples are also rebuilt for every commit and are available as artifacts from the CI jobs. I haven't wanted to make too much of a deal over them because I‌ don't think the examples are that great yet. Eventually I'd like to see them getting posted to the website gallery, but the basic workflow is here already. 

**Not** addressed points are:

* The drop cap situation, see #394. My chance I happened on a line width metric that isn't currently blowing up quite as bad as it was when you opened this issue, but it is still not fixed right. The leading after the drop cap is still finicky and the wrapping in the paragraph gets all goofed up until the end of that paragraph (what caused the space after "InDesign" and "before "Let's"). This is likely a pushback bug, but I'm leaving the drop-cap issue to track that. 

* The showoff document is still not (in my opinion) all that compelling. See #343.

* The line break tolerance hacks here shouldn't be needed, the out of the box performance shouldn't be so bad on narrow columns. This is left to #620‌ to track.

AS for issues with the manual as mentioned by @pkulchenko:

* TOC number alignment has been fixed, see #859.
* Layout around floating images is related to the drop-cap issue, but also tracked in #860.
* Big empty space in the manual is tracked in #890.

Thanks everybody for your input on this. Sorry it has taken so long to address.

**Comment by pkulchenko** [comment]
Thank you for the update @alerque! It's great to see that progress is being made and these issues are fixed or being fixed.
